src/control/server/storage/bdev/firmware.go

```diff
@@ -147,7 +165,58 @@ func getDeviceController(pciAddr string, controllers storage.NvmeControllers) (*
 		}
 	}
 
-	return nil, fmt.Errorf("no NVMe controller found with PCI address %q", pciAddr)
+	return nil, FaultPCIAddrNotFound(pciAddr)
+}
+
+func filterControllers(controllers storage.NvmeControllers, modelID, fwRev string) storage.NvmeControllers {
+	selected := make(storage.NvmeControllers, 0, len(controllers))
+	for _, ctrlr := range controllers {
+		if filterMatches(modelID, ctrlr.Model) &&
+			filterMatches(fwRev, ctrlr.FwRev) {
+			selected = append(selected, ctrlr)
+		}
+	}
+	return selected
+}
+
+func filterMatches(filterStr, actualStr string) bool {
+	return filterStr == "" || strings.ToUpper(actualStr) == strings.ToUpper(filterStr)
+}
+
+// UpdateFirmware updates the NVMe device controller firmware.
+func (p *Provider) UpdateFirmware(req FirmwareUpdateRequest) (*FirmwareUpdateResponse, error) {
+	if p.shouldForward(req) {
+		return p.fwFwd.Update(req)
+	}
+
+	if len(req.FirmwarePath) == 0 {
+		return nil, errors.New("missing path to firmware file")
+	}
+
+	controllers, err := p.getRequestedControllers(req.DeviceAddrs, req.ModelID, req.FirmwareRev, false)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(controllers) == 0 {
+		return nil, errors.New("no NVMe device controllers")
+	}
+
+	resp := &FirmwareUpdateResponse{
+		Results: make([]DeviceFirmwareUpdateResult, 0, len(controllers)),
+	}
+	for _, con := range controllers {
+		err = p.backend.UpdateFirmware(con.PciAddr, req.FirmwarePath, defaultFirmwareSlot)
+		result := DeviceFirmwareUpdateResult{
+			Device: *con,
+		}
+		if err != nil {
+			result.Error = err.Error()
+		}
+		resp.Results = append(resp.Results, result)
+	}
```

Comment:
This is fine, but you could also set the length and omit capacity. Since you're setting up a slice of actual structs rather than pointers-to-struct, the empty structs will be created for you and you can just use the index from range to get at them.

e.g.

```go
	resp := &FirmwareUpdateResponse{
		Results: make([]DeviceFirmwareUpdateResult, len(controllers)),
	}
	for i, con := range controllers {
		err = p.backend.UpdateFirmware(con.PciAddr, req.FirmwarePath, defaultFirmwareSlot)

                resp.Results[i].Device = *con
                if err != nil {
                        resp.Results[i].Error = err.Error()
                }
	}
```

Don't know that one way or the other is better, but wanted to point this out as a possibility.

Reply:
Your alternate seems a bit less verbose so I'm switching over to it. Thanks!